rules-engine/src/rules_engine/engine.py

```diff
@@ -71,73 +49,137 @@ def average_indoor_temp(
-    initial_bp -- Balance point temperature in degrees F to try at first
-    bp_sensitivity -- Degrees F to add/subtract from balance point when refining
-    """
-
-    # this implementation assumes that the three list arguments are the same
-    # length, and also that the fuel_type string is valid. Not sure if this is
-    # the place to do that checking, or if it will be handled earlier
-
-    bills_days = [len(l) for l in avg_temps_lists]
-    avg_daily_heating_usages = [
-        usage / days - non_heat_usage for usage, days in zip(usages, bills_days)
-    ]
-    period_hdds = [period_hdd(temps, initial_bp) for temps in avg_temps_lists]
-
-    btu_per_usage = {
-        "gas": 100000,  # usage in therms
-        "oil": 139600,  # usage in gallons
-        "propane": 91333,  # usage in gallons
-    }
-    bpu = btu_per_usage[fuel_type]
-
-    uas = [
-        ua(
-            bills_days[period],
-            avg_daily_heating_usages[period],
-            bpu,
-            heat_sys_efficiency,
-            period_hdds[period],
+class FuelType(Enum):
+    """Enum for fuel types. Values are BTU per usage"""
+
+    GAS = 100000
+    OIL = 139600
+    PROPANE = 91333
+
+
+class Home:
+    """Defines attributes and methods for calculating home heat metrics"""
+
+    def __init__(
+        self,
+        fuel_type: FuelType,
+        heat_sys_efficiency: float,
+        initial_balance_point: float = 60,
+        thermostat_set_point: float = 68,
+    ):
+        self.fuel_type = fuel_type
+        self.heat_sys_efficiency = heat_sys_efficiency
+        self.balance_point = initial_balance_point
+        self.thermostat_set_point = thermostat_set_point
+
+    def initialize_billing_periods(
+        self,
+        temps: List[List[float]],
+        usages: List[float],
+        avg_non_heating_usage: float = 0,
+    ) -> None:
+        """Eventually, this method should categorize the billing periods by
+        season and calculate avg_non_heating_usage based on that. For now, we
+        just pass in winter-only heating periods and manually define non-heating
+        """
+        # assume for now that temps and usages have the same number of elements
+        self.bills = []
+        self.avg_non_heating_usage = avg_non_heating_usage
+
+        for i in range(len(usages)):
+            self.bills.append(BillingPeriod(temps[i], usages[i], self))
+
+    def calculate_balance_point_and_ua(
+        self, balance_point_sensitivity: float = 2
+    ) -> None:
+        """Calculates the estimated balance point and UA coefficient for the home,
+        removing UA outliers based on a normalized standard deviation threshold.
+        """
+        self.uas = [bp.ua for bp in self.bills]
+        self.avg_ua = sts.mean(self.uas)
+        self.stdev_pct = sts.pstdev(self.uas) / self.avg_ua
+        self.refine_balance_point(balance_point_sensitivity)
+
+        while self.stdev_pct > 0.10:
```

Comment:
Should we make these sensitivities/magic numbers configurable constants?

Reply:
Latest commit makes 0.1 into a parameter